### PR TITLE
CellLogicProtocol

### PR DESCRIPTION
**Pull request by MaksymMelnyk3**
Try Create CellLogic Protocol

### CODE REVIEW

File: Sources/WebSocketManager/WebSocketManager.swift [hunk 1]
```diff
@@ -47,26 +50,43 @@ public class WebSocketManager {
...
+                self.playerAction(gameData)
             default:
                 print("raw")
             }
         }
     }
     
+    private func playerAction(_ data: GameDataModel) {
+        let result = self.cellLogic.handleGameUpdate(mapState: data)
+        let myData = PlayerAction(key: "player_action", data: PlayerData(cells: [PlayerCell(id: (result?.myCells.first?.cellId ?? ""), velocity: PlayerVelocity(x: (result?.myCells.first?.velocity?.x ?? 0), y: (result?.myCells.first?.velocity?.y ?? 0)), speed: (result?.myCells.first?.speed ?? 0), growIntention: PlayerGrowIntention(eatEfficiency: (result?.myCells.first?.growIntention?.eatEfficiency ?? 0), maxSpeed: (result?.myCells.first?.growIntention?.maxSpeed ?? 0), power: (result?.myCells.first?.growIntention?.power ?? 0), mass: (result?.myCells.first?.growIntention?.mass ?? 0), volatilization: (result?.myCells.first?.growIntention?.volatilization ?? 0)))]))
```
Comment (spanastasia) [inline Sources/WebSocketManager/WebSocketManager.swift]: We need to send data for all user cells as we aim to support split and merge functionality.

File: Sources/WebSocketManager/Models/testModels/Velocity.swift [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+//    }
+//
+//    public func moveToPosition(target: Position) -> Velocity {
+//        let fromX = self.x
+//        let fromY = self.y
+//        let targetX = target.x!
+//        let targetY = target.y!
+//        return Velocity(x: targetX - fromX!, y: targetY - fromY!)
+//    }
+//}
```
Comment (spanastasia) [inline Sources/WebSocketManager/Models/testModels/Velocity.swift]: Do you need this commented code?

File: Sources/WebSocketManager/Models/GameDataModel.swift [hunk 3]
```diff
@@ -7,42 +7,81 @@
...
+    public let canMerge: Bool?
+    public let position: Position?
+    public let own: Bool?
 }
 
-struct Direction: Codable {
-    let x, y: Double?
+//public struct Direction: Codable {
+//    public let x, y: Double?
+//}
```
Comment (spanastasia) [inline Sources/WebSocketManager/Models/GameDataModel.swift]: Commented code

File: Sources/WebSocketManager/Models/testModels/CellActivity.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
+//
+//  File.swift
+//  
+//
+//  Created by Maxim on 14.07.2022.
+//
+
+import Foundation
+
+public struct CellActivity {
```
Comment (spanastasia) [inline Sources/WebSocketManager/Models/testModels/CellActivity.swift]: Looks like CellActivity might conform Encodable protocol, just to pass array of CellActivity to JSONEncoder

File: Sources/WebSocketManager/Models/GameDataModel.swift [hunk 5]
```diff
@@ -7,42 +7,85 @@
...
+public struct Velocity: Codable {
+    public var x: Double?
+    public var y: Double?
+    
+    public init (x: Double, y: Double){
+        self.x = x
+        self.y = y
+    }
+
+    public func moveToPosition(target: Position) -> Velocity {
```
Comment (ghost) [inline Sources/WebSocketManager/Models/GameDataModel.swift]: This function should be in Cell Class

**Comment by ghost** [comment]
Need to implement the logic for caching the game state and retrieving data from the cache if it is not received from the backend.

**Pull request merged**